src/AutoSchedule.cpp

```diff
@@ -0,0 +1,2312 @@
+#include<algorithm>
+
+#include "AutoSchedule.h"
+
+namespace Halide {
+namespace Internal {
+
+using std::string;
+using std::vector;
+using std::map;
+using std::set;
+using std::deque;
+using std::pair;
+using std::make_pair;
+
+void simplify_box(Box &b) {
+    for (size_t i = 0; i < b.size(); i++) {
+        b[i].min = simplify(b[i].min);
+        b[i].max = simplify(b[i].max);
+    }
+}
+
+struct FStage {
+    Function func;
+    uint32_t stage_num;
+    FStage(Function _func, uint32_t _stage_num) :
```

Comment:
nit: leading underscopre in var/arg names is not halide style. it's ok to duplicate the arg and member names here, there's no ambiguity